module/VuFind/src/VuFind/View/Helper/Root/AbstractClassBasedTemplateRenderer.php

```diff
@@ -42,25 +43,33 @@
 abstract class AbstractClassBasedTemplateRenderer extends AbstractHelper
 {
     /**
-     * Recursively locate and render a template that matches the provided class
-     * name (or one of its parent classes); throw an exception if no match is
-     * found.
+     * Cache for found templates
      *
-     * @param string $template     Template path (with %s as class name placeholder)
-     * @param string $className    Name of class to apply to template.
-     * @param string $topClassName Top-level parent class of $className (or null
-     * if $className is already the top level; used for recursion only).
+     * @var array
+     */
+    protected $templateCache = [];
+
+    /**
+     * Recursively locate a template that matches the provided class name
+     * (or one of its parent classes); throw an exception if no match is found.
+     *
+     * @param string            $template     Template path (with %s as class name
+     * placeholder)
+     * @param string            $className    Name of class to apply to template.
+     * @param ResolverInterface $resolver     Resolver to use
+     * @param string            $topClassName Top-level parent class of $className
+     * (or null if $className is already the top level; used for recursion only).
      *
      * @return string
      * @throws RuntimeException
      */
     protected function resolveClassTemplate($template, $className,
```

Comment:
Changing the signature and return content of this method potentially introduces a backward compatibility break, which makes me wonder if we should defer this PR until 8.0. I thought about whether we could make the same changes without the BC break, and while we could, I think what you've done actually makes the method naming more accurate, so I don't think it makes sense to go out of our way to prevent change.

Of course, if we are already introducing a BC break, maybe an even better thing to do is to move this functionality into a Trait and deprecate this abstract base class, since the Trait approach is more flexible and cleaner.

Reply:
I agree that 8.0 makes sense as the target. I moved the functionality to a trait. I kept it in View/Helper/Root since it assumes AbstractHelper with getView() as the base class. It could be made more generic by adding the view helper as a parameter, but I couldn't think of a need outside view helpers and ensuring that helps keep the semantics sane. Do you agree?

Reply:
Yes, this makes sense to me. It might not hurt to add a comment indicating that the trait should only be used in a view helper because of the dependency on getView(), but leaving it in this namespace may be enough to imply that! In any case, this looks good to me, so I'll just leave the PR open for now, and will retest/merge when we're ready to start 8.0 development in the dev branch.

Reply:
Thanks, I added a short comment at the top of the trait.